Python/compile.c

```diff
@@ -6042,3 +6053,311 @@ PyAST_Compile(mod_ty mod, const char *filename, PyCompilerFlags *flags,
+                        break;
+                }
+                break;
+
+            case JUMP_ABSOLUTE:
+            case JUMP_FORWARD:
+                switch(target->i_opcode) {
+                    case JUMP_FORWARD:
+                        inst->i_target = target->i_target;
+                        break;
+                    case JUMP_ABSOLUTE:
+                    case RETURN_VALUE:
+                    case RERAISE:
+                    case RAISE_VARARGS:
+                        lineno = inst->i_lineno;
+                        *inst = *target;
+                        inst->i_lineno = lineno;
+                        break;
+                }
+                break;
+        }
+    }
+    return 0;
+error:
+    return -1;
+}
+
+
+static void
+clean_basic_block(basicblock *bb) {
+    /* Remove NOPs and any code following a return or re-raise. */
+    int dest = 0;
+    for (int src = 0; src < bb->b_iused; src++) {
+        switch(bb->b_instr[src].i_opcode) {
+            case NOP:
+                /* skip */
+                break;
+            case RETURN_VALUE:
+            case RERAISE:
+                bb->b_next = NULL;
+                bb->b_instr[dest] = bb->b_instr[src];
+                dest++;
+                goto end;
+            default:
+                if (dest != src) {
+                    bb->b_instr[dest] = bb->b_instr[src];
+                }
+                dest++;
+                break;
+        }
+    }
+end:
+    assert(dest <= bb->b_iused);
+    bb->b_iused = dest;
+}
+
+static int
+mark_reachable(struct assembler *a) {
+    basicblock **stack, **sp;
+    sp = stack = (basicblock **)PyObject_Malloc(sizeof(basicblock *) * a->a_nblocks);
+    if (stack == NULL) {
+        return -1;
+    }
+    basicblock *entry = a->a_reverse_postorder[0];
+    entry->b_reachable = 1;
+    *sp++ = entry;
+    while (sp > stack) {
+        basicblock *b = *(--sp);
+        if (b->b_next && b->b_next->b_reachable == 0) {
+            b->b_next->b_reachable = 1;
+            *sp++ = b->b_next;
+        }
+        for (int i = 0; i < b->b_iused; i++) {
+            basicblock *target;
+            if (b->b_instr[i].i_jrel || b->b_instr[i].i_jabs) {
+                target = b->b_instr[i].i_target;
+                if (target->b_reachable == 0) {
+                    target->b_reachable = 1;
+                    *sp++ = target;
+                }
+            }
+        }
+    }
+    return 0;
+}
+
+
+/* Perform basic peephole optimizations on a control flow graph.
+   The consts object should still be in list form to allow new constants
+   to be appended.
+
+   All transformations keep the code size the same or smaller.
+   For those that reduce size, the gaps are initially filled with
+   NOPs.  Later those NOPs are removed.
+*/
+
+static int
+optimize_cfg(struct assembler *a, PyObject *consts)
+{
+    for (int i = 0; i < a->a_nblocks; i++) {
+        if (optimize_basic_block(a->a_reverse_postorder[i], consts)) {
```

Comment:
Does this converge on one single iteration? What happens if you have a sequence of many jumps all together?

Reply:
No. Some jumps that can be eliminated will still remain.

This PR is aimed at closely mimicking the  behavior of the existing peephole optimizer. 
We can improve it later.